plutus-ledger/src/Ledger/Tx/CardanoAPI.hs

```diff
@@ -480,52 +496,51 @@ toCardanoTxInWitness
         <*> pure zeroExecutionUnits
         )
 
-toCardanoMintWitness :: PV1.Redeemers -> Int -> P.MintingPolicy -> Either ToCardanoError (C.ScriptWitness C.WitCtxMint C.BabbageEra)
+toCardanoMintWitness :: Redeemers -> Int -> P.MintingPolicy -> Either ToCardanoError (C.ScriptWitness C.WitCtxMint C.BabbageEra)
 toCardanoMintWitness redeemers idx (P.MintingPolicy script) = do
-    let redeemerPtr = PV1.RedeemerPtr PV1.Mint (fromIntegral idx)
+    let redeemerPtr = RedeemerPtr Mint (fromIntegral idx)
     P.Redeemer redeemer <- maybe (Left MissingMintingPolicyRedeemer) Right (Map.lookup redeemerPtr redeemers)
     C.PlutusScriptWitness C.PlutusScriptV1InBabbage C.PlutusScriptV1
         <$> fmap C.PScript (toCardanoPlutusScript (C.AsPlutusScript C.AsPlutusScriptV1) script)
         <*> pure C.NoScriptDatumForMint
-        <*> pure (C.fromPlutusData $ PV1.toData redeemer)
+        <*> pure (C.fromPlutusData $ toData redeemer)
         <*> pure zeroExecutionUnits
 
--- TODO Handle reference script once 'P.TxOut' supports it (or when we use
--- exclusively 'C.TxOut' in all the codebase).
-fromCardanoTxOut :: C.TxOut C.CtxTx era -> Either FromCardanoError PV1.TxOut
-fromCardanoTxOut (C.TxOut addr value datumHash _) =
-    PV1.TxOut
+fromCardanoTxOut :: C.TxOut C.CtxTx era -> Either FromCardanoError TxOut
+fromCardanoTxOut (C.TxOut addr value outputDatum refScript) =
+    TxOut
     <$> fromCardanoAddressInEra addr
     <*> pure (fromCardanoTxOutValue value)
-    <*> pure (fromCardanoTxOutDatumHash datumHash)
+    <*> pure (fromCardanoTxOutDatum outputDatum)
+    <*> pure (fromCardanoReferenceScript refScript)
+
+fromCardanoReferenceScript :: C.ReferenceScript era -> Maybe ScriptHash
+fromCardanoReferenceScript (C.ReferenceScript _rtisidsie sial) = fmap scriptHash $ fromCardanoScriptInAnyLang sial
+fromCardanoReferenceScript C.ReferenceScriptNone               = Nothing
 
 toCardanoTxOut
     :: C.NetworkId
-    -> (Maybe P.DatumHash -> Either ToCardanoError (C.TxOutDatum ctx C.BabbageEra))
-    -> PV1.TxOut
+    -> (OutputDatum -> Either ToCardanoError (C.TxOutDatum ctx C.BabbageEra))
+    -> (Maybe ScriptHash -> Either ToCardanoError (C.ReferenceScript C.BabbageEra))
+    -> TxOut
     -> Either ToCardanoError (C.TxOut ctx C.BabbageEra)
-toCardanoTxOut networkId fromHash (PV1.TxOut addr value datumHash) =
+toCardanoTxOut networkId fromOutputDatum fromScriptHash (TxOut addr value outputDatum refScript) =
```

Comment:
I am extending the mechanism for datum lookups to the reference scripts, hoping we have a solution for it.